### DISCUSSION TITLE
STM32F4 Discovery LIS3DSH Accelerometer via SPI

### DISCUSSION THREAD

**Discussion by joshegrimes** [zephyrproject-rtos/zephyr#73810] [Q&A] [answered]
Hi, I am currently trying to work with the LIS3DSH Accelerometer via SPI on an STM32F4 Discovery kit and am running into some issues likely with the Zephyr driver support? (this is my best guess) Essentially the issue is that when using the existing LIS2DH Zephyr drivers, we can't get past a device_is_ready check. I am relatively new and inexperienced in Zephyr (college student here), but my colleague suggested it was likely a driver compatibility issue with the sensor. I would try to debug/fix this issue myself, but I wouldn't even know where to start in terms of how to debug/alter driver code...

So, I guess my question is if anyone has done SPI communication with this sensor before and could lead me in the right direction, or has any guidance on how to debug? Code is shown below.

main.c 

```
/*
 * STM32F4DISCO Accelerometer Testing
 * Josh Grimes
 * Summer 2024
 */

#include <stdio.h>
#include <zephyr/kernel.h>
#include <zephyr/device.h>
#include <zephyr/drivers/sensor.h>
#include <zephyr/drivers/spi.h>

/*
void fetch_and_display(const struct device *accel)
{
	struct sensor_value acceleration[3];
	int rc = sensor_sample_fetch(accel);
	rc = sensor_channel_get(accel, SENSOR_CHAN_ACCEL_XYZ, acceleration);
	printf("x %f, y %f, z %f", sensor_value_to_double(&acceleration[0]), sensor_value_to_double(&acceleration[1]), sensor_value_to_double(&acceleration[2]));
}
*/

int main(void)
{
	const struct device *const accel = DEVICE_DT_GET(DT_ALIAS(lis3dsh));

	if (accel == NULL) 
	{
		printf("No device found\n");
		return 0;
	}
	if (!device_is_ready(accel)) 
	{
		printf("Device %s is NOT ready\n", accel->name);
		return 0;
	}
	else
	{
		printf("Device %s IS ready\n", accel->name);
	}
	
	/*
	while(1)
	{
		fetch_and_display(accel);
	}
	*/

	return 0;
}
```
(the code to get sensor data and display admittedly may be very wrong currently - just trying to ensure the sensor is ready first)

stm32f4_disco.overlay
```
/ {
	aliases 
    {
		lis3dsh = &acc_lis3dsh_spi;
	};
};


&spi1
{
    status="okay";
    pinctrl-0 = <&spi1_sck_pa5 &spi1_miso_pa6 &spi1_mosi_pa7>;
    pinctrl-names = "default";
    cs-gpios = <&gpioe 3 GPIO_ACTIVE_LOW>;

    acc_lis3dsh_spi: lis3dsh@0
    {
            status="okay";
            compatible = "st,lis3dsh", "st,lis2dh", "st,lis3dh";
            reg = <0>;
            spi-max-frequency = <10000000>;
    };
};
```
prj.conf
```
CONFIG_SPI=y
CONFIG_SPI_ASYNC=y
CONFIG_I2C=n
CONFIG_GPIO=y
CONFIG_SPI_STM32=y
CONFIG_PINCTRL=y

CONFIG_CONSOLE=y
CONFIG_UART_CONSOLE=y
CONFIG_SERIAL=y
CONFIG_STDOUT_CONSOLE=y
CONFIG_CBPRINTF_FP_SUPPORT=y
CONFIG_PRINTK=y

CONFIG_SENSOR=y
CONFIG_LIS2DH=y
CONFIG_LIS2DH_TRIGGER_GLOBAL_THREAD=y
```
(again, new to Zephyr so a lot of this is probably redundant, but I was trying different things in order to debug)

My output on the serial monitor looks like this:
```
---- Closed the serial port COM9 ----
---- Opened the serial port COM9 ----
*** Booting Zephyr OS build v3.6.0 ***
Device lis3dsh@0 is NOT ready
```

Also, I am just using Zephyr's given stm32f4_disco.dts for my board .dts file, in case that matters (didn't make a custom one or anything).

One thing I did notice that further suggested a lack of support was that on the Bindings Index page [(page)](https://docs.zephyrproject.org/latest/build/dts/api/bindings.html#dt-vendor-st) there is a st,lis2dh (on i2c bus), a st,lis2dh (on spi bus), and a st,lis3dh (on i2c bus) listed, but no "st,lis3dh (on spi bus)" like I would expect there to be... So maybe this is all a moot point, but thought someone might have better knowledge of this than my team and I.

Thanks in advance!

**Comment by ajarmouni-st** [comment]
Hi @joshegrimes, I will try to answer some of your questions

> LIS3DSH Accelerometer via SPI

It appears that LIS3DSH is not at all supported in Zephyr (there is no compatible devicetree binding), potentially it can be used with the same driver as LIS2DH & LIS3DH , just need to add this binding in `dts/bindings/sensor/st,lis3dsh-spi.yaml`:
```
# Copyright (c) 2024 STMicroelectronics
# SPDX-License-Identifier: Apache-2.0

description: STMicroelectronics LIS3DSH 3-axis "nano" accelerometer accessed through SPI bus

compatible: "st,lis3dsh"

include: ["spi-device.yaml", "st,lis2dh-common.yaml"]
```
but I am not sure (@avisconti If you could share your input, thanks!).

> how to debug driver code

You should first add `CONFIG_SENSOR_LOG_LEVEL_DBG=y` to your `prj.conf`

> CONFIG_SPI_ASYNC=y

If you want to use SPI in async mode, you should have `CONFIG_SPI_STM32_INTERRUPT=y`as well.

To work with sensors in Zephyr, you need `CONFIG_SENSOR=y`.

**Comment by avisconti** [comment, marked as answer]
Yes, I agree with @ajarmouni-st.
Basilically, despite the part number name which is simliar, LIS3DSH has NOTHING to do with lis2dh/lis3dh/lis3dh12. In fact the device id is 0x3F for the first one, while it is 0x33 for the other three which belong to the same family and are register compatible (in this case you can use lis2dh driver for all three devices). So, LIS3DSH is currently not supported in zephyr and it will never be, as it is declared obsolete (see https://www.st.com/content/st_com/en/search.html#q=lis3dsh-t=products-page=1)

**Comment by avisconti** [comment]
As LIS3DSH is not supported and obsolete, if you really require to use that device I suggest you to add the support by your own, starting from lis2dh driver. The minimum required is the driver and dts bindings.

**Comment by ajarmouni-st** [comment]
@joshegrimes This can be helpful https://github.com/STMicroelectronics/lis3dsh-pid/tree/master

**Comment by joshegrimes** [comment]
I see, thanks for all the helpful information! I may try to add support myself at some point, but right now it's not a huge priority for my team and I... more just wanted to know for sanity's sake. I will update if I ever get it to work...

Thanks again!
